### PR TITLE
fix: hasPlugin does not track parent plugins

### LINKED ISSUE #4925: `hasPlugin` does not check outer scopes
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.20.0

### Plugin version

_No response_

### Node.js version

18.16.1

### Operating system

Windows

### Operating system version (i.e. 20.04, 11.3, 10)

11

### Description

`hasPlugin` does not check outer scopes for registered plugins

### Steps to Reproduce

```
import fastify from 'fastify'
import fastifyPlugin from 'fastify-plugin'

const server = fastify()

const pluginCallback = (server, options, done) => {
    done()
}
const pluginName = 'awesome-plugin'
const plugin = fastifyPlugin(pluginCallback, {name: pluginName})

await server.register(plugin)

await server.register(async (server) => {
    const hasPlugin = server.hasPlugin(pluginName)
    console.log(hasPlugin) // prints: false

    const hasPluginWorkaround =
    server[Symbol.for('registered-plugin')]
        .includes(pluginName)
    console.log(hasPluginWorkaround) // prints: true
})
```

### Expected Behavior

`hasPlugin` should be `true`

### PR DESCRIPTION
**Pull request by Eomm**
closes #4925

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/pluginOverride.js [hunk 1]
```diff
@@ -48,8 +49,14 @@ module.exports = function override (old, fn, opts) {
   instance[kSchemaController] = SchemaController.buildSchemaController(old[kSchemaController])
   instance.getSchema = instance[kSchemaController].getSchema.bind(instance[kSchemaController])
   instance.getSchemas = instance[kSchemaController].getSchemas.bind(instance[kSchemaController])
-  instance[pluginUtils.registeredPlugins] = Object.create(instance[pluginUtils.registeredPlugins])
-  instance[kPluginNameChain] = [pluginUtils.getPluginName(fn) || pluginUtils.getFuncPreview(fn)]
+
+  // Track the registered and loaded plugins since the root instance.
+  // It does not track the current encapsulated plugin.
+  instance[pluginUtils.kRegisteredPlugins] = Object.create(instance[pluginUtils.kRegisteredPlugins])
```
Comment (Uzlopak) [inline lib/pluginOverride.js:55]: Shouldnt kRegisteredPlugins be in symbols.js?

**Review by metcoder95 — APPROVED** [review]
LGTM 👍

**Review by mcollina — APPROVED** [review]
lgtm

However I feel this _might_ have unwanted consequences and it might be safer to land it in `next`. Wdyt?

**Comment by Fdawgs** [comment]
Agree with @mcollina, definitely a breaking change.

File: lib/pluginOverride.js (hunk 1 above)
Comment (Eomm) [reply lib/pluginOverride.js:55]: I though the same 👍🏼

Comment (Uzlopak) [reply lib/pluginOverride.js:55]: Will you move it to there?

**Comment by Eomm** [comment]
> Maybe we could just land it and see if it breaks anybody and revert thereafter?

I think so 👍🏼

**Comment by mcollina** [comment]
I'm ok with that approach.
Anybody objects?

**Review by Uzlopak — APPROVED** [review]
LGTM

**Pull request merged**